supernode.cpp

```diff
@@ -1,28 +1,60 @@
 #include <stdexcept>
 #include <cmath>
 #include <boost/multiprecision/cpp_int.hpp>
-#include <omp.h>
 #include "include/supernode.h"
 #include "include/util.h"
 #include "include/graph_worker.h"
 
-Supernode::Supernode(uint64_t n, long seed): sketches(log2(n)), idx(0), logn(log2
-(n)), seed(seed), n(n) {
+uint32_t Supernode::bytes_size;
+
+Supernode::Supernode(uint64_t n, long seed): idx(0), logn(log2(n)),
+					     n(n), seed(seed), sketch_size(Sketch::sketchSizeof()) {
+
   // generate logn sketches for each supernode (read: node)
-  for (int i=0;i<logn;++i)
-    sketches[i] = new Sketch(n*n, seed++);
+  for (int i = 0; i < logn; ++i) {
+    Sketch::makeSketch(get_sketch(i), seed++);
+  }
+}
+
+Supernode::Supernode(uint64_t n, long seed, std::fstream &binary_in) :
+  idx(0), logn(log2(n)), n(n), seed(seed), sketch_size(Sketch::sketchSizeof()) {
+
+  // read logn sketches from file for each supernode (read: node)
+  for (int i = 0; i < logn; ++i) {
+    Sketch::makeSketch(get_sketch(i), seed++, binary_in);
+  }
+}
+
+Supernode::SupernodeUniquePtr Supernode::makeSupernode(uint64_t n, long seed) {
+  Sketch::n = n * n;
+  set_size(n);
+
+  void *loc = malloc(bytes_size);
+  SupernodeUniquePtr ret(new (loc) Supernode(n, seed), [](Supernode* s){ free(s); });
+  return ret;
+}
+
+Supernode::SupernodeUniquePtr Supernode::makeSupernode(uint64_t n, long seed, std::fstream &binary_in) {
+  Sketch::n = n * n;
+  set_size(n);
+
+  void *loc = malloc(bytes_size);
+  SupernodeUniquePtr ret(new (loc) Supernode(n, seed, binary_in), [](Supernode* s){ free(s); });
+  return ret;
+}
+
+Supernode* Supernode::makeSupernode(void* loc, uint64_t n, long seed) {
+  return new (loc) Supernode(n, seed);
 }
 
 Supernode::~Supernode() {
-  for (int i=0;i<logn;++i)
```

Comment:
can you remind me why we don't need to worry about de-allocating?

Reply:
Because the memory is contiguous and is freed by the graph code at the top.

Reply:
So the calls to delete are redundant

Reply:
Also because they're probably unique pointers